### ISSUE TITLE
Hard deadlines

### ISSUE THREAD

**Issue by espenak** [devilry/devilry-django#287]
We should support **hard** deadlines. This should be optional, and off by default. Users need to be made aware of the consequences of using hard deadlines when they choose the option.
## Sponsored by

This issue is requested and payed for by [University of Oslo, Department of Informatics](http://www.mn.uio.no/ifi/).
## Target branch

We should implement this as part of the `subjectadmin` branch , which is going to be merged into master very soon (see #288).  The changes will have to be made in both the new and the old admin interface. The userinterface implementation details below refer to the new subjectadmin interface.
## Overview of the consequences of hard deadlines

(we need to clean this up before putting it in the userinterface)
- Students with valid reasons for not making the deadline (sick etc.) will have to ask someone to move their deadline (or add a new deadline). With soft deadlines the students can add their delivery, and the examiner can clearly see that they delivered after the deadline.
- Students have to deliver before the deadline, and it will be impossible to add deliveries on an expired deadline.
- Hard deadlines cause extra work for both students and administrators/examiners if Devilry is unavailable (system maintainance, power outage, ...) at the time of the deadline because:
  - Someone must move the deadline, or create a new deadline.
  - Students need to be notified about the new deadline, and then re-try adding a delivery. This can be a huge problem for home exams because it is hard to require that students add deliveries after the deadline (they may have other plans than waiting for the Devilry to become available), and the extra wait for an admin to create/move a deadline can easily become several days of delay (E.g.: weekends), which gives the "late" students extra time.
## User interface implementation details
- A checkbox under "advanced options" when creating an assignment.
- Clearly stated (with an edit button) in the assignment overview.
  - Something like this should be shown in the sidebar:
    ### Deadline handling
    
    Hard - Students can **not** add deliveries when the deadline has expired _(edit | more info)_
    
    _or_
    ### Deadline handling
    
    Soft - Students **can add** deliveries when the deadline has expired _(edit | more info)_
  - The _more info_ button should show the advantage/disadvantage of soft/hard deadlines:
    - A cleaned version of the consequences listed above, and the reverse for soft deadlines.
- Make it impossible to add deliveries in the web-interface if `hard_deadlines==True`:
  - Basically the same as the message we are currently displaying when the deadline has expired, however we also hide the user interface components used to upload files, and show a different message.
## Core implementation details
- An `int` field, `deadline_handling`, on the assignment model.
- Check for `hard_deadlines` when adding deliveries, and throw exception if deadline is expired and `deadline_handling==HARD`.
